lib/modules/bulk_data/test/authorization_sequence_test.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+require_relative '../../../../test/test_helper'
+
+describe Inferno::Sequence::BulkDataAuthorizationSequence do
+  before do
+    @sequence_class = Inferno::Sequence::BulkDataAuthorizationSequence
+    config = load_json_fixture(:bulk_data_authorization)
+
+    @instance = Inferno::Models::TestingInstance.create(
+      url: 'http://www.example.com',
+      bulk_client_id: config['client_id'],
+      bulk_public_key: config['public_key'].to_json,
+      bulk_private_key: config['private_key'].to_json,
+      bulk_token_endpoint: config['token_url']
+    )
+
+    @client = FHIR::Client.new(@instance.url)
+
+    @access_token = {
+      'token_type' => 'bearer',
+      'expires_in' => 900,
+      'access_token' => 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ0b2tlbl90eXBlIjoiYmVhcmVyIiwiZXhwaXJlc19pbiI6OTAwLCJpYXQiOjE1NzM5NDU1MDQsImV4cCI6MTU3Mzk0NjQwNH0.Ds-9HxQPJshkPYYBowJXltTaX2T6MSv_qYnZLjteTH8',
+      'scope' => 'system/*.read'
+
+    }
+  end
+
+  def self.it_test_rquired_parameter(test_id, request_headers: nil)
+    before do
+      @test = @sequence_class[test_id]
+      @sequence = @sequence_class.new(@instance, @client)
+    end
+
+    it 'pass with stastus code 400' do
+      a_request = stub_request(:post, @instance.bulk_token_endpoint)
```

Comment:
When inferno is making requests, I think it is important that we test the request as well as the response. This stub doesn't verify that each test's request is malformed in the way that we expect.

Reply:
Malformed requests are tested in test sequence 01-11.

Reply:
Yes, but this stub doesn't verify that the requests we send are actually malformed.

Reply:
Can you clarify? We are not going to send malformed request in test 12.

Reply:
Oh. I was looking at the wrong line. Because the request doesn't matter here. What we are testing here is that if the server returns 400, the test sequence should handle 400 correctly.

Reply:
Yes, but we also need to test that Inferno is making the requests correctly.

These tests basically have two parts:
- Make a request in a particular way
- Expect a particular type of response

The unit tests are currently returning particular types of responses and verifying that the test passes or fails. This is good, but it only checks half of the tests behavior.

The unit tests are not checking whether each test makes its request in the particular way it needs to. For example, one test needs to omit `grant_type`, another needs to have an invalid `iss`. If the unit tests don't check that the requests match these requirements, then at any time in the future the code could change such that the requests are no longer being made in the correct way without causing the unit tests to fail. The way to check these requests is to call `stub_request` separately for each unit test so that it will only match against the correct type of request rather than the current behavior where it matches any POST to the correct endpoint.

Reply:
The client_claim is unique for each request since both expires_in and jti change very time a claim is created. So the claim generated from test sequence and test stub are different though parameters having same values. It is possible to inject those values into the test sequence as attribute.

Reply:
It doesn't have to be an exact match. You should be able to use a block in webmock to inspect the body and verify that the portion of the jwt that this test cares about is correct: https://github.com/bblimke/webmock#matching-requests-against-provided-block

Reply:
Have solved most problem but the correct_key test is still a headache since the private key used could not be retrieved/tested from the payload

Reply:
That is for test 11? Would it be reasonable to check that the signature can't be validated with the expected public key?

Reply:
I was thinking how to test that it uses a specific private key. You are right. I only need to test if it not uses the correct key.

Reply:
The logic is so complicated. I think I need another set of unit tests to test these unit tests